flink-formats/flink-parquet/src/test/java/org/apache/flink/formats/parquet/ParquetTableSourceTest.java

```diff
@@ -0,0 +1,230 @@
+import java.util.List;
+
+import static org.junit.Assert.assertArrayEquals;
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertFalse;
+import static org.junit.Assert.assertNotNull;
+import static org.junit.Assert.assertTrue;
+import static org.mockito.Mockito.any;
+import static org.mockito.Mockito.doReturn;
+import static org.mockito.Mockito.eq;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.spy;
+import static org.mockito.Mockito.times;
+import static org.mockito.Mockito.verify;
+import static org.mockito.Mockito.when;
+
+
+/**
+ * Test cases for {@link ParquetTableSource}.
+ */
+public class ParquetTableSourceTest extends TestUtil {
+	private static final AvroSchemaConverter SCHEMA_CONVERTER = new AvroSchemaConverter();
+
+	@ClassRule
+	public static TemporaryFolder tempRoot = new TemporaryFolder();
+
+	@Test
+	public void testGetReturnType() throws Exception {
+		Tuple3<Class<? extends SpecificRecord>, SpecificRecord, Row> nested = TestUtil.getNestedRecordTestData();
+		Path path = TestUtil.createTempParquetFile(tempRoot.getRoot(), TestUtil.NESTED_SCHEMA,
+			Collections.singletonList(nested.f1));
+
+		MessageType nestedSchema = SCHEMA_CONVERTER.convert(TestUtil.NESTED_SCHEMA);
+		ParquetTableSource parquetTableSource = ParquetTableSource.builder()
+			.path(path.getPath())
+			.forParquetSchema(nestedSchema)
+			.build();
+
+		TypeInformation<Row> returnType = parquetTableSource.getReturnType();
+		assertNotNull(returnType);
+		assertTrue(returnType instanceof RowTypeInfo);
+		RowTypeInfo rowType = (RowTypeInfo) returnType;
+		assertEquals(NESTED_ROW_TYPE, rowType);
+	}
+
+	@Test
+	public void testGetTableSchema() throws Exception {
+		Tuple3<Class<? extends SpecificRecord>, SpecificRecord, Row> nested = TestUtil.getNestedRecordTestData();
+		Path path = TestUtil.createTempParquetFile(tempRoot.getRoot(), TestUtil.NESTED_SCHEMA,
+			Collections.singletonList(nested.f1));
+
+		MessageType nestedSchema = SCHEMA_CONVERTER.convert(TestUtil.NESTED_SCHEMA);
+		ParquetTableSource parquetTableSource = ParquetTableSource.builder()
+			.path(path.getPath())
+			.forParquetSchema(nestedSchema)
+			.build();
+
+		TableSchema schema = parquetTableSource.getTableSchema();
+		assertNotNull(schema);
+
+		RowTypeInfo expectedSchema = (RowTypeInfo) NESTED_ROW_TYPE;
+		assertArrayEquals(expectedSchema.getFieldNames(), schema.getFieldNames());
+		assertArrayEquals(expectedSchema.getFieldTypes(), schema.getFieldTypes());
+	}
+
+	@Test
+	public void testFieldsProjection() throws Exception {
+		ParquetTableSource parquetTableSource = createNestedTestParquetTableSource();
+
+		ParquetTableSource projected = (ParquetTableSource) parquetTableSource.projectFields(new int[] {2, 4, 6});
+
+		// ensure copy is returned
+		assertTrue(projected != parquetTableSource);
+
+		// ensure table schema is the same
+		assertEquals(parquetTableSource.getTableSchema(), projected.getTableSchema());
+
```

Comment:
Check that the source description is different (`!parquetTableSource.explainSource.equals(projected.explainSource())`)